### PR TITLE
[WIP] Executables define rule for C object files (-output-obj)

### LINKED ISSUE #22: Support for -output-obj
Jbuilder seems not yet able to create standalone shared library `-output-obj`. However it is not very different from linking executables (you just need a +PIC switch). What would be the right way to add it to jbuilder?

 - a new stanza?
 - an option to executables?

### PR DESCRIPTION
**Pull request by bobot**
Should fix #22

Remains:
 * [ ] select the right extension .so/.dll
 * [ ] disable the rule if not a compiler with `+PIC`? or a better error messages?

### CODE REVIEW

**Comment by diml** [comment]
You can use anything from `Context.t`, in particular you can use `ctx.ext_dll` :)

Then, so that we can reference this shared object file without hardcoding `.so` in jbuild files, you should also add an `ext_dll` variable in `dollar_var_map`.

We shouldn't disable the rule, instead we should use:

```ocaml
Build.fail { fail = fun () -> die "blah blah this is not supported" }
```

So that we get a nice failure when requesting to build this target.

**Comment by diml** [comment]
I'm sure fPIC is actually the right thing to look at. For instance I can see that there is a libasmrun_shared.so, which I assume is used for `-output-obj`. Maybe it's only needed for `-output-complete-obj` and/or only when you have C stubs. For pure OCaml code, I assume that it wouldn't be an issue unless `-nodynlink` is used.

/cc @mshinwell who has been using it and might know more.

If it's too complicated, let's just always enable the rule for now. It's only going to be a problem if you want to optionally install such a `.so` with opam. But I assume nobody installs such a `.so` in opam currently?

BTW there is the question of -output-obj vs -output-complete-obj. I don't know if it make sense to generate both at once. If it does maybe we could have `foo.so` and `foo.complete.so`

**Comment by whitequark** [comment]
> I'm sure fPIC is actually the right thing to look at. For instance I can see that there is a libasmrun_shared.so, which I assume is used for -output-obj. Maybe it's only needed for -output-complete-obj and/or only when you have C stubs. For pure OCaml code, I assume that it wouldn't be an issue unless -nodynlink is used.

This is not correct. There are three runtimes:
  * libasmrun.a, which is built without -fPIC,
  * libasmrun_pic.a, which is built with -fPIC,
  * libasmrun_shared.so, which is libasmrun_pic.a linked as a shared object.

I do not know what is the purpose of libasmrun_shared.so, and I do not believe the compiler will use it anywhere by itself (though you could specify it with -runtime-variant if you so wish).

Any of the runtimes can be used with either `-output-obj` and `-output-complete-obj`. You want to use the PIC ones when the output of `-output*-obj` is linked into a shared object, and non-PIC ones otherwise.

As for the difference between -output-obj and -output-complete-obj, the latter links in the C stubs from autolink libraries but the former does not. Arguably -output-obj should have always done that, but I tried it and I think it was a compatibility hazard. In general, no one should use -output-obj anymore.

**Comment by diml** [comment]
@whitequark thanks for the clarifications.

Let's use only `-output-complete-obj` then. I hadn't though about it but it's true that you can also produce a `.o` with `-output*-obj`. In the end we have a lot of possible combinations: `{byte,native}x{executable,object,shared-object}` and I'm not sure it's worth systematically setting up all the rules given that they'll rarely be used.

`(library ...)` stanzas have a `(modes ...)` field. We should add something similar for `(executalbes ...)` stanzas but with a few more modes: `{byte,native}_{object,shared_object}`. The default would be `(byte native)` as for libraries.

@bobot, BTW there is now a `test` directory with sandboxed tests. Can you add a test for this feature?

**Comment by diml** [comment]
@bobot I think we'll often forget to update the timestamp file. In practice, we should only need to rebuild boot.exe when the jbuild format evolves and we use the new feature in the jbuild files of jbuilder, so probably not that often. If this is a persistent problem, we could try to add `{bin,src,test,...}/jbuild` as dependencies of `boot.exe` in the `Makefile`. That should do the trick

**Comment by wokalski** [comment]
This is important when it comes to cross compilation. With cross compilation I want to be able to produce an object file which would be linked to the final executable.

**Comment by diml** [comment]
@bobot, sorry i missed your last commits... I'll review this ASAP

**Comment by diml** [comment]
Ok, yh that seems complicated instead. Happy to scale back. So technically we just remove the `Object` mode and keep only `Executable` and `Shared_object`, is that what you have in mind?

Another thought: currently, if the user wants to do themselves one of the mode that is not supported, it's really painful to construct the right command line. Maybe what we could do is the following: add a `custom_linkages` field in `executables` stanzas:

```scheme
(executable
 (...
  (custom_linkages
   ((mode native)
    (extension ${ext_obj})
    (flags (--output-obj))))))
```

then, if a common pattern appears we can add builtin support for it.

What do you think?

**Comment by wokalski** [comment]
Sorry for chiming in but static libraries are useful in certain situations. For instance in my use case I can’t use sharedl ibraries because of Apple’s restrictions. They are supported - yes but it’s not allowed to use them in an App Store app.

**Comment by diml** [comment]
@wokalski would the `custom_linkages` idea work for you?

I imagine that you would write:

```
(executable
 (name foo)
 (custom_linkages
  ((mode native)
   (extension ${ext_obj})
   (flags (--output-obj)))))
```

it's a bit verbose, but it would unblock the situation until it's clearer what to do

**Comment by wokalski** [comment]
Yes, it would. I’m sorry; I missed your message.

**Comment by diml** [comment]
@bobot, IIUC, `-output-obj` could be used as an optimization when building for several different runtime variants. So we could use it inside Jbuilder to implement all this but I don't  think it's necessary to  expose it to the user.

By the way I'm getting a lot of errors `Cannot overwrite existing file /tmp/camlobj...` with OCaml 4.05. Looking at the compiler code, it seems to me that `-output-complete-obj` cannot be used with OCaml < 4.06, do you agree?

**Comment by diml** [comment]
Ok, the doc still need to be updated but I think the code is ready.

In the end we have the following: the `(modes ...)` field of executable stanzas now defines what linking modes will be used. A linking mode is a pair `(<mode> <kind>)` where `<mode>` is either `byte` or `native` and `<kind>` is one of `exe`, `object` and `shared_object`. And we can use the following short-hands: 

- `exe` or `native` for `(native exe)`
- `object` for `(native object)`
- `shared_object` for `(native object)`
- `byte` for `(byte exe)`

The following extensions are used for each linking mode:

mode | kind | ext
-------|------|-----
byte | exe | .bc
native | exe | .exe
byte | object | .bc${ext_obj}
native | object | .exe${exe_obj}
byte | shared_object | .bc${ext_dll}
native | shared_object | ${ext_dll}

## Issues

when compiling a `.so` with `ocamlopt -ouput-complete-obj`, `-ldl -lm` is not passed to the C compiler and as a result we get an error about the math symbols when dynamically loading the `.so`. To workaround this issue Dune is manually passing `-ccopt X` for every `X` in `native_c_libraries`.

I also checked the Jane Street jenga rules, and I can see that we are passing `-ccopt -Wl,-z,now` when compiling a .so with either ocamlc or ocamlopt with the following comment:

```ocaml
        (* side-step issue where the compiler expects register r10 to be unchanged across
           a call to caml_call_gc@plt, which is not true if the symbol resolution happens
           right then. This works by forcing the symbol resolution to happen at load time
           instead.  *)
```

File: src/exe.ml [hunk 1]
```diff
@@ -45,6 +45,52 @@ module Linkage = struct
...
+  let so_flags = ["-output-complete-obj"; "-runtime-variant"; "_pic"]
+
+
+  let of_user_config (ctx : Context.t) (m : Jbuild.Executables.Link_mode.t) =
+    let has_native = Option.is_some ctx.ocamlopt in
+    let mode = m.mode in
+    match m.kind with
+    | Exe ->
+      if mode = Native && not has_native then
+        custom
```
[the remark this answers is not in the document]
Comment (diml) [reply src/exe.ml]: Yes. I'm not completely sure about this, but it seems simpler overall.

**Comment by rgrinberg** [comment]
I think the pr can be merged without this, buy why is this hard to fix?
Can't we just filter the linkages for what's available and then die if that
list is empty.

**Comment by diml** [comment]
Basically this is the problem: currently, if you don't put a `modes` field, then there will always be a rule for  `.exe` that will be native if possible, and bytecode if not.

For shared objects, the rule are not setup by default. So one needs a `modes` field. However, there is no way to get the `native or byte` behavior. There are several choices for this problem:

- we can systematically setup the rules for the various combination, and with a `modes` field, the choice will be `native or byte`
- we add a mode `best` to represent `native or byte`

Thinking about it, the 2nd solution seems best.

**Comment by diml** [comment]
BTW, @whitequark, does the following issue rings a bell: `ocamlc -output-complete-obj ...` links with the C libraries listed in `ocamlc -config | grep bytecomp_c_libraries` but `ocamlopt -output-complete-obj ...` doesn't link with the C libraries listed in `ocamlopt -config | grep native_c_libraries`?

**Comment by diml** [comment]
@rgrinberg I have added the doc. Could you review the last changes?

**Comment by rgrinberg** [comment]
OK, looks good. Can we have a test for a mode that isn't available if we have a good way to hide ocamlopt in a test?

**Comment by diml** [comment]
Sure, we have some tests for this in `byte-code-only`, I added a test of an error case. The error message is not great, ideally we should still setup the rules but make them error out with a nice error message saying that native compilation is not available.

**Pull request merged**